src/rgw/driver/rados/rgw_notify.cc

```diff
@@ -1201,23 +1224,34 @@ reservation_t::reservation_t(const DoutPrefixProvider* _dpp,
 }
 
 reservation_t::reservation_t(const DoutPrefixProvider* _dpp,
-			     rgw::sal::RadosStore* _store,
-			     rgw::sal::Object* _object,
-			     rgw::sal::Object* _src_object,
-			     rgw::sal::Bucket* _bucket,
-			     const std::string& _user_id,
-			     const std::string& _user_tenant,
-			     const std::string& _req_id,
-			     optional_yield y) :
-    dpp(_dpp), store(_store), s(nullptr), size(0) /* XXX */,
-    object(_object), src_object(_src_object), bucket(_bucket),
-    object_name(nullptr),
-    metadata_fetched_from_attributes(false),
-    user_id(_user_id),
-    user_tenant(_user_tenant),
-    req_id(_req_id),
-    yield(y)
-{}
+                             rgw::sal::RadosStore* _store,
+                             rgw::sal::Object* _object,
+                             rgw::sal::Object* _src_object,
+                             rgw::sal::Bucket* _bucket,
+                             const std::string& _user_id,
+                             const std::string& _user_tenant,
+                             const std::string& _req_id,
+                             optional_yield y)
+    : dpp(_dpp),
+      store(_store),
+      s(nullptr),
+      size(0) /* XXX */,
+      object(_object),
+      src_object(_src_object),
+      bucket(_bucket),
+      object_name(nullptr),
+      metadata_fetched_from_attributes(false),
+      user_id(_user_id),
+      user_tenant(_user_tenant),
+      req_id(_req_id),
+      yield(y) {
+  if (store->svc()->zone->get_zonegroup().supports(
```

Comment:
can we not do that in the dtor?
at which point do we need to query the attribute?

Reply:
> can we not do that in the dtor?
> at which point do we need to query the attribute?

Destructor you mean ?
Naah we need it during the publish reserve, which is called right after constructing this structure

Reply:
but "publish_reserve" has a return value that you can use to indicate failure, while the ctor does not have

Reply:
> but "publish_reserve" has a return value that you can use to indicate failure, while the ctor does not have

I mentioned earlier, publish-reserve is common for s3 and non s3 both. And we need to load the attrs only for non-s3. It's an additional rados call for s3 op. And this function is in HOT path, as it's called frequently, so we could avoid calling it in publish-reserve. 
And as said earlier, if bucket-reload was a problem in ctor, it already would be bigger problem and code will not even reach up to publish-reserve

Reply:
you are right. but there is an indication that could eb checked in `publish_reserve()`. see: https://github.com/ceph/ceph/pull/54868#discussion_r1434039351